cirq/neutral_atoms/neutral_atom_devices.py

```diff
@@ -0,0 +1,421 @@
+                             ops.CCXPowGate,
+                             ops.CCZPowGate)):
+            if not gate.exponent.is_integer():
+                raise ValueError('controlled gates must have integer exponents')
+
+    def validate_operation(self, operation: ops.Operation):
+        """
+        Raises an error if the given operation is invalid on this device.
+
+        Args:
+            operation: the operation to validate
+
+        Raises:
+            ValueError: If the operation is not valid
+        """
+        if not isinstance(operation, (ops.GateOperation,
+                                      ops.ParallelGateOperation)):
+            raise ValueError('Unsupported operation: {!r}'.format(operation))
+
+        # The gate must be valid
+        self.validate_gate(operation.gate)
+
+        # All qubits the operation acts on must be on the device
+        for q in operation.qubits:
+            if q not in self.qubits:
+                raise ValueError('Qubit not on device: {!r}'.format(q))
+
+        if isinstance(operation.gate, (ops.MeasurementGate, ops.IdentityGate)):
+            return
+
+        # Verify that a controlled gate operation is valid
+        if isinstance(operation, ops.GateOperation):
+            if len(operation.qubits) > self._max_parallel_c:
+                raise ValueError("Too many qubits acted on in parallel by a "
+                                 "controlled gate operation")
+            if len(operation.qubits) > 1:
+                for p in operation.qubits:
+                    for q in operation.qubits:
+                        if self.distance(p, q) > self._control_radius:
+                            raise ValueError("Qubits {!r}, {!r} are too "
+                                             "far away".format(p, q))
+            return
+
+        # Verify that a valid number of Z gates are applied in parallel
+        if isinstance(operation.gate, ops.ZPowGate):
+            if len(operation.qubits) > self._max_parallel_z:
+                raise ValueError("Too many Z gates in parallel")
+
+        # Verify that a valid number of XY gates are applied in parallel
+        if isinstance(operation.gate, (ops.XPowGate,
+                                       ops.YPowGate,
+                                       ops.PhasedXPowGate)):
+            if (len(operation.qubits) > self._max_parallel_xy and
+                    len(operation.qubits) != len(self.qubits)):
+                raise ValueError("Bad number of XY gates in parallel")
+
+    def validate_moment(self, moment: ops.Moment):
+        """
+        Raises an error if the given moment is invalid on this device
+
+        Args:
+            moment: The moment to validate
+
+        Raises:
+            ValueError: If the given moment is invalid
+        """
+        super().validate_moment(moment)
+
+        GATE_CATEGORIES = {
+            ops.ZPowGate: 'Z',
+            ops.pauli_gates._PauliZ: 'Z',
```

Comment:
Hmmmmmm... it is not a good sign that we had to list these. Brittle.

Perhaps what we need instead is:

```python
CATEGORIES = {'Z': [...], 'XY': [...], ...}
for op in operations:
    for k, v in CATEGORIES.entries():
        if isinstance(op, v):
            categorized_ops[k].append(op)
```

Reply:
I like this better since isinstance plays much nicer with inheritance than type. I've implemented this.